### PR TITLE
fix(clerk-js): Avoid triggering prepare verification twice

### PR DESCRIPTION
**Pull request by panteliselef**
## Description
Backport of #2459 

### Before

https://github.com/clerk/javascript/assets/19269911/5a94d861-be69-4007-addd-bb239924f5b3

### After

https://github.com/clerk/javascript/assets/19269911/5af85b9d-ea13-42c5-8b03-31d007278664

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/backend`
- [ ] `@clerk/chrome-extension`
- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/fastify`
- [ ] `gatsby-plugin-clerk`
- [ ] `@clerk/localizations`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/remix`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/themes`
- [ ] `@clerk/types`
- [ ] `build/tooling/chore`

### CODE REVIEW

**Comment by panteliselef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @panteliselef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/chrome-extension | 0.6.3-snapshot.vd68f379 |
| @clerk/clerk-js | 4.68.2-snapshot.vd68f379 |
| @clerk/clerk-expo | 0.19.28-snapshot.vd68f379 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@0.6.3-snapshot.vd68f379 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@4.68.2-snapshot.vd68f379 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@0.19.28-snapshot.vd68f379 --save-exact
```

**Comment by panteliselef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @panteliselef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/chrome-extension | 0.6.3-snapshot.v88ff6e0 |
| @clerk/clerk-js | 4.68.2-snapshot.v88ff6e0 |
| @clerk/clerk-expo | 0.19.28-snapshot.v88ff6e0 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@0.6.3-snapshot.v88ff6e0 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@4.68.2-snapshot.v88ff6e0 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@0.19.28-snapshot.v88ff6e0 --save-exact
```

File: packages/clerk-js/src/ui/components/SignUp/SignUpPhoneCodeCard.tsx [hunk 1]
```diff
@@ -1,19 +1,21 @@
...
 export const SignUpPhoneCodeCard = withCardStateProvider(() => {
   const signUp = useCoreSignUp();
 
-  React.useEffect(() => {
-    // TODO: This prepare method is not idempotent.
-    // We need to make sure that R18 won't trigger this twice
-    void prepare();
+  useEffectForceOnce(() => console.log('NICE'));
+  useEffect(() => {
+    console.log('nice');
```
Comment (nikosdouvlis) [inline packages/clerk-js/src/ui/components/SignUp/SignUpPhoneCodeCard.tsx:14]: Nice :)

**Review by nikosdouvlis — APPROVED** [review]
This needs to be solved in the resource level as usage of utils like `useEffectForceOnce` is generally discouraged, but I'm happy to merge it as a hotfix for v4.

@panteliselef can you make sure we don't have a similar hook already? I remember us using something like this in 2022 when React 18 came out. Not sure if we still use it anywhere however

**Comment by panteliselef** [comment]
@nikosdouvlis don't merge yet, this is not resolved yet. The best approach would be to either introduce a `useMutation` similar to `useFetch` or allow cancellation in our clerk-js methods for resources (Abort Controller)

**Comment by panteliselef** [comment]
!snapshot

File: packages/clerk-js/src/ui/components/SignUp/SignUpPhoneCodeCard.tsx (hunk 1 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/SignUp/SignUpPhoneCodeCard.tsx:15]: I'm debugging on another app that is not local in my machine and i need the logs + snapshots :)

**Comment by clerk-cookie** [comment]
Hey @panteliselef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/chrome-extension | 0.6.3-snapshot.v42e17ec |
| @clerk/clerk-js | 4.68.2-snapshot.v42e17ec |
| @clerk/clerk-expo | 0.19.28-snapshot.v42e17ec |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@0.6.3-snapshot.v42e17ec --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@4.68.2-snapshot.v42e17ec --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@0.19.28-snapshot.v42e17ec --save-exact
```

**Comment by panteliselef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @panteliselef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/chrome-extension | 0.6.3-snapshot.vaf7adfd |
| @clerk/clerk-js | 4.68.2-snapshot.vaf7adfd |
| @clerk/clerk-expo | 0.19.28-snapshot.vaf7adfd |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@0.6.3-snapshot.vaf7adfd --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@4.68.2-snapshot.vaf7adfd --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@0.19.28-snapshot.vaf7adfd --save-exact
```

**Comment by panteliselef** [comment]
!snapshot

**Comment by panteliselef** [comment]
!snapshot

**Comment by panteliselef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @panteliselef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/chrome-extension | 0.6.3-snapshot.v6ca6305 |
| @clerk/clerk-js | 4.68.2-snapshot.v6ca6305 |
| @clerk/clerk-expo | 0.19.28-snapshot.v6ca6305 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@0.6.3-snapshot.v6ca6305 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@4.68.2-snapshot.v6ca6305 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@0.19.28-snapshot.v6ca6305 --save-exact
```

**Comment by panteliselef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @panteliselef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/chrome-extension | 0.6.3-snapshot.v8fcca55 |
| @clerk/clerk-js | 4.68.2-snapshot.v8fcca55 |
| @clerk/clerk-expo | 0.19.28-snapshot.v8fcca55 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@0.6.3-snapshot.v8fcca55 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@4.68.2-snapshot.v8fcca55 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@0.19.28-snapshot.v8fcca55 --save-exact
```

**Comment by panteliselef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @panteliselef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/chrome-extension | 0.6.3-snapshot.va439d8e |
| @clerk/clerk-js | 4.68.2-snapshot.va439d8e |
| @clerk/clerk-expo | 0.19.28-snapshot.va439d8e |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@0.6.3-snapshot.va439d8e --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@4.68.2-snapshot.va439d8e --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@0.19.28-snapshot.va439d8e --save-exact
```

**Comment by panteliselef** [comment]
@nikosdouvlis Do u mind taking another look ? Changes have been made since the first push

**Comment by leonardonelson91** [comment]
image

I'm having this issue until this day

Next.JS 14.2
Clerk next: 4.29.12

**Pull request merged**